presto-spi/src/main/java/com/facebook/presto/spi/function/ComplexTypeFunctionDescriptorFunctionDescriptor.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.common.Subfield.allSubscripts;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptySet;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Contains properties that describe how the function operates on Map or Array inputs.
+ */
+public class ComplexTypeFunctionDescriptorFunctionDescriptor
+{
+    public static final ComplexTypeFunctionDescriptorFunctionDescriptor DEFAULT = new ComplexTypeFunctionDescriptorFunctionDescriptor(
+            true,
+            true,
+            true,
+            emptyList(),
+            Optional.of(emptySet()),
+            Optional.empty());
+
+    /**
+     * Indicates whether the function internally accessing subfields. This does not include the subfield access through the functional parameter (lambda) passed as an
+     * argument. If function accepts the functional argument that can access the subfields, then the <code>lambdaArgumentIndex</code> must contain the index of such
+     * functional argument and <code>lambdaArgumentToCallArgumentIndexMap</code> must contain the mapping between the lambda argument to the index of the argument of the
+     * CallExpression that this argument relates to.
+     */
+    private final boolean isAccessingInputValues;
+
+    /**
+     * Indicates whether the semantic of the function transforms the shape of the input.
+     * If function does not transform the shape of the input, then all collected subfields accessed from lambdas from outer calls could be passed to the input column or next
```

Comment:
Added to PushdownSubfields 
```
if (functionDescriptor.isAccessingInputValues()
    || (functionDescriptor.isTransformingInput()) && 
    !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
    // If function internally accesses the input values we cannot prune any unaccessed subfields since we do not know what subfields function accessed.
    // If function transforms the shape of the input map or array argument and the form of such transformation was not supplied in
    // getOutputToInputTransformationFunction() then we have to discard all collected accessed subfields in lambdas of outer calls because we do not know how
    // to attribute those subfields to the input argument. Additionally, we have to stop collecting lambda subfields until we find another function that
    // makes pruning subfields possible.
```